src/encode.ts

```diff
@@ -0,0 +1,477 @@
+	redactErrors?: boolean;
+};
+
+export function encode(
+	value: unknown,
+	{ plugins = [], redactErrors = true }: EncodeOptions = {},
+) {
+	return new ReadableStream<string>({
+		async start(controller) {
+			let refCache = new WeakMap();
+			let asyncCache = new WeakMap();
+			let counters = { refId: 0, promiseId: 0 };
+			let wg = new WaitGroup();
+			let chunks: string[] = [];
+
+			let encode = (value: unknown) => {
+				encodeSync(
+					value,
+					chunks,
+					refCache,
+					asyncCache,
+					promises,
+					counters,
+					plugins,
+					redactErrors,
+				);
+
+				controller.enqueue(chunks.join("") + "\n");
+				chunks.length = 0;
+			};
+
+			let handlePromiseResolved = (id: number, value: unknown) => {
+				wg.done();
+				controller.enqueue(`${id}${STR_SUCCESS}`);
+				encode(value);
+			};
+
+			let handlePromiseRejected = (id: number, error: unknown) => {
+				wg.done();
+				controller.enqueue(`${id}${STR_FAILURE}`);
+				encode(error);
+			};
+
+			let promises = {
+				push: (...promiseFrames: AsyncFrame[]) => {
+					for (let [type, id, promise] of promiseFrames) {
+						wg.add();
+						if (type === 1) {
+							(promise as PromiseLike<unknown>).then(
+								handlePromiseResolved.bind(null, id),
+								handlePromiseRejected.bind(null, id),
+							);
+						} else {
+							(async () => {
+								let iterator = (promise as AsyncIterable<unknown>)[
+									Symbol.asyncIterator
+								]();
+
+								let result: IteratorResult<unknown>;
+								do {
+									result = await iterator.next();
+									if (!result.done) {
+										controller.enqueue(`${id}${STR_SUCCESS}`);
+										encode(result.value);
+									}
+								} while (!result.done);
+							})()
+								.then(() => {
+									controller.enqueue(`${id}\n`);
+								})
+								.catch((error) => {
+									controller.enqueue(`${id}${STR_FAILURE}`);
+									encode(error);
+								})
+								.finally(() => {
+									wg.done();
+								});
+						}
+					}
+				},
+			};
+
+			try {
+				encode(value);
+
+				do {
+					await wg.wait();
+				} while (wg.p > 0);
+
+				controller.close();
+			} catch (error) {
+				controller.error(error);
+			}
+		},
+	});
+}
+
+type EncodeFrameObj =
+	| { type: 1; prefix: string; value: unknown }
+	| { type: 2; prefix: string; value: undefined };
+
+class EncodeFrame {
+	public type: number;
+	public prefix: string;
+	public value: unknown;
+	constructor(type: number, prefix: string, value: unknown) {
+		this.type = type;
+		this.prefix = prefix;
+		this.value = value;
+	}
+}
+
+export function encodeSync(
+	value: unknown,
+	chunks: { push(...chunk: string[]): void },
+	refs: WeakMap<object, number>,
+	promises: WeakMap<object, number>,
+	asyncFrames: { push(frame: AsyncFrame): void },
+	counters: { refId: number; promiseId: number },
+	plugins: EncodePlugin[],
+	redactErrors: boolean,
+) {
+	let encodeStack: EncodeFrameObj[] = [
+		new EncodeFrame(1, "", value) as EncodeFrameObj,
```

Comment:
As `esbuild` will inline enums anyways (at least according to https://github.com/evanw/esbuild/issues/128#issuecomment-1002341244), I would suggest using a `const enum` here instead of magic  numbers for `type` - might make maintenance a bit easier over the years.

Reply:
I'm going for TS features that can be stripped instead of transpiled.

Reply:
The cast here is just for additional type safety BTW (it's easier to type the shape here than the constructor signature). Classes just ended up being more efficient at runtime.